src/pyodide-py/pyodide/webloop.py

```diff
@@ -0,0 +1,429 @@
+#! /usr/bin/env python
+# -*- coding: utf-8 -*-
+# flake8: noqa
+
+"""
+
+A custom event loop for running asyncio in Pyodide
+
+Version: 0.1.2
+
+Author: Wei OUYANG (@oeway)
+        Adapted from the EventSimulator made by @damonjw
+        https://gist.github.com/damonjw/35aac361ca5d313ee9bf79e00261f4ea
+
+Version History:
+        * support contextvars
+        * add event loop policy
+        * throw error when the loop is already running
+        * support Python 3.8
+        * fix run_until_complete
+"""
+
+import heapq
+import asyncio
+import time
+import traceback
+import contextvars
+
+import js
+from typing import Dict, Tuple, Optional, Awaitable, Callable
+
+
+class WebLoop(asyncio.AbstractEventLoop):
+    """A custom event loop for running asyncio in Pyodide
+
+    It works by utilizing the browser event loop via the setTimeout function
+    """
+
+    def __init__(self, debug: Optional[bool] = False, interval: Optional[int] = 10):
+        """
+        Instatiate the web loop
+        """
+        self._running = False
+        self._immediate = []
+        self._scheduled = []
+        self._futures = []
+        self._next_handle = None
+        self._debug = debug
+        self._stop = False
+        self._interval = interval
+        self._timeout_promise = js.eval(
+            "self._timeoutPromise = function(time){return new Promise((resolve)=>{setTimeout(resolve, time);});}"
+        )
+        self._exception_handler = None
+        self._task_factory = self._default_task_factory
+        self._stop_callbacks = None
+        self._result = None
+        self._exception = None
+
+    def get_debug(self):
+        """
+        Get the debug mode (bool) of the event loop.
+        """
+        return self._debug
+
+    def set_debug(self, enabled: bool):
+        """
+        Set the debug mode of the event loop.
+        """
+        self._debug = enabled
+
+    def time(self):
+        """
+        Return the current time, as a float value, according to the time module (time.time()).
+        """
+        return time.time()
+
+    def run_forever(self):
+        """
+        Run the event loop until stop() is called.
+
+        Note that this function is different from the standard asyncio loop implementation in two ways:
+         1) It won't block the execution
```

Comment:
Could you elaborate on why that's not the issue when using it and the motivation behind it.

Reply:
This is a fun one. It's actually impossible for it to block: the majority of async calls that need to be done require the help of the browser, in particular they run on the browser event loop. If our duplicate event loop blocks, the browser event loop never gets a chance to run and the IO we're waiting for never happens. (Also in native python we would go over to a different thread to schedule tasks because the one thread is blocked. This is hard without threading.)

Reply:
So does this mean that this is essentially a busy loop? (ie constantly running a tight loop until a something changes)

Reply:
@itszn I think it sort of means the opposite of that. With a busy loop, we would starve out the browser event loop. Normal native Python event loops use `select` to wait on other threads. We basically just tell the browser event loop to do the work and then wait on it.

By the way, in my opinion this PR is superseded by #1158 where I took the ideas here and simplified the implementation as much as possible.

Reply:
(This is just for my own understanding)
As far as I was aware, in browsers a WASM export can't block (This is why sleep is a busy loop in emscriptn
https://emscripten.org/docs/porting/pthreads.html#blocking-on-the-main-browser-thread). How does the python event loop wait for the js resolve without returning out of wasm? I know one option is asyncify but that isn't supported yet right?

Reply:
This is all pretty confusing, it took me a good deal of careful contemplation to make sense of it.

> in browsers a WASM export can't block

That's correct. 

> How does the python event loop wait for the js resolve without returning out of wasm?

We do return out of wasm, this is necessary because if we don't return from wasm we block the browser event loop and nothing gets resolved. We use setTimeout. The main loop is:
```python
def _do_tasks(self, ...):
    # do various stuff
    setTimeout(self._interval, self._do_tasks)
```

Similarly, in PR #1158 I define `call_later` to directly do setTimout:
https://github.com/iodide-project/pyodide/blob/a1c30a9329d8fd6265ecbf2b915b0aa175b79e81/src/pyodide-py/pyodide/simple_web_loop.py#L95

> I know one option is asyncify but that isn't supported yet right?

asyncify is to allow normal synchronous C functions to do async calls. I think the primary point of this is that in native C you can do IO by blocking the current thread until the OS gets back to you with the result. This can't work in browser because there is only one thread and if that thread blocks no IO ever happens.
I believe the way this works is that emscripten saves the current stack and yields to the browser event loop, then when the asynchronous task is finished emscripten "rewinds" the stack and then continues as if nothing happened.

In our case we don't need anything like asyncify because the python APIs we're using are already asynchronous: an async coroutine is a type of generator, when you await a future, it yields the future to the event loop. Then when the future is resolved, the event loop calls `coroutine.send(result_of_future)`, and the execution continues from the `await` keyword with `result_of_future` as the result of the `await`.

Reply:
I'd suggest you try experimenting with `send` directly. Consider the following, which you can do in 
https://pyodide-cdn2.iodide.io/dev/full/console.html

```python
import asyncio
class DumbLoop(asyncio.AbstractEventLoop):    
   def create_future(self):
      return asyncio.Future(loop=self)
   def get_debug(self):
      return False
asyncio.set_event_loop(DumbLoop())
from js import fetch
async def test():
   resp = await fetch("packages.json")
   json = await resp.json()
   return json
c = test()
```
Now `c` is a coroutine. If you run `c.send(None)` that starts the coroutine running and runs it up to the first `await`.
```python
fut1 = c.send(None)
```
Now `fut1` is a `Future` which wraps the `Promise` returned by `fetch`. It will resolve when the `fetch` finishes. Wait a few seconds, then:
```python
fut1.result() # ==> [object Response]
```
The response object looks like:
```javascript
Response { type: "basic", url: "https://pyodide-cdn2.iodide.io/dev/full/packages.json", redirected: false, status: 200, ok: true, statusText: "OK", headers: Headers, body: ReadableStream, bodyUsed: false }
```
anyways now that we have the result, the coroutine is expecting us to send it in:
```python
fut2 = c.send(fut1.result())
```
the coroutine stores the response object `fut1.result()` into `resp` and then runs until `await resp.json()`. The return value `fut2` is a future that resolves when `resp.json()` is done.
Now `fut2.result()` will be the actual json object, deserialized into javascript. Finally, we're supposed to send this back into the coroutine. The coroutine stores it into `json` and then `return json` has the effect of `raise StopIteration(json)`:
```python
try:
   c.send(fut2.result())
except StopIteration as retval:
   result = retval
```
Now `result` is the json object.